### ISSUE TITLE
Doesn't work with jquery.formset.js

### ISSUE THREAD

**Issue by datakid** [applegrew/django-select2#109]
When used in a formset, in conjunction with jquery.formset.js to add new formsets, the js isn't loaded on the new formset?

jquery.formset.js has this function:

   ``added``
        If you set this to a function, that function will be called each
        time a new form is added. The function should take a single argument,
        ``row``; it will be passed a jQuery object, wrapping the form that
        was just added.

What part of django-select2 should I call to activate this? I note that you mention that some work needs to be done once the html is created?
https://github.com/applegrew/django-select2/issues/90#issuecomment-32058767

**Comment by witchnick** [comment]
I believe Rizumu has done a lot of work with dynamic inline formsets integration. However after few hours of thinking I'm still stuck and I do not know how to use it. I use  jquery.formset.js and I think I should use django_select2() JS plugin but really I do not know how to do it :( 
Please give me some hints or prepare an example with it. I think it should be also included in Demoproject.

**Comment by rizumu** [comment]
@codingjoe I'm unable to get dynamic formsets working with the new codebase. Do you have a suggestion on what could be done in the `added` callback?

**Comment by codingjoe** [comment]
@anneFly, you did some work on that if I remember correctly. How did you solve for our code?

**Comment by anneFly** [comment]
@rizumu you have to initialize the select2 field manually afte it was added to the DOM. So the callback of dynamic formsets to do what [this code](https://github.com/applegrew/django-select2/blob/master/django_select2/static/django_select2/django_select2.js) does. If it's not a heavy field, it's simple:
```
added: function (row) {
    row.find('.django-select2').not('django-select2-heavy').select2();
}
```
for heave fields you need to duplicate the whole logic for heavy fields:
```
added: function (row) {
    row.find('.django-select2.django-select2-heavy').each(function () {
        var field_id = $(this).data('field_id');
        $(this).select2({
            ajax: {
                data: function (params) {
                    return {
                        term: params.term,
                        page: params.page,
                        field_id: field_id
                    }
                },
                processResults: function (data, page) {
                    return {
                        results: data.results,
                        pagination: {
                            more: data.more
                        }
                  }
                }
            }
        });
    });
}
```
(I didn't test this code but something like this should work)

@codingjoe the PR I did was basically wrapping this code into a jQuery plugin, so that the code can be reused directly and doesn't have to be duplicated.

**Comment by codingjoe** [comment]
@anneFly, I'm still for merging the pull-requests. If you add an example to the documentation, I could even write a selenium test myself.
Thanks

**Comment by rizumu** [comment]
Thanks @anneFly and @codingjoe. I am using heavy widgets here and I've been trying the same as you suggest the last few days. After I click 'add another' I then see two select2 inputs on the page, one that partially works (I can search the ajax results but selecting is wonky) and the other still in the uninitialised state.

Could it be the formset library, is there a suggested one that is known to work? I switched from django-superformset to django-dynamic-formset with no luck. I also tried out https://github.com/witchnick/jquery.formset.js-for-django-select2/blob/master/jquery.formset.js 

Another issue I see, which may or may not be related, is that when the widget is cloned by any of these formset libraries, the same pickled widget_id/field_id is used for the cloned inlines. 

It would be helpful to see a working example, with a dynamic formset library and heavy data.

**Comment by anneFly** [comment]
@rizumu it was a bit tricky, but I made a working example of select2 and django-dynamic-formset.
See my repository [here](https://github.com/anneFly/django-dynamic-formset-select2-poc). Look at the html and js in the [template](https://github.com/anneFly/django-dynamic-formset-select2-poc/blob/master/poc/app/templates/books.html). The tricky part was actually [this line](https://github.com/anneFly/django-dynamic-formset-select2-poc/blob/master/poc/app/templates/books.html#L43). I hope this will help you.

**Comment by codingjoe** [comment]
@anneFly :+1: :100:

**Comment by rizumu** [comment]
Thank you @anneFly this is very helpful. In the meantime, I discovered that `django-dynamic-formset` was the root cause of the issue, because switching to this fork ` https://github.com/adongy/django-dynamic-formset` and the added and deleted methods also to active the fields. I didn't dig into the exact fix that this fork implements, but would like to when I have some more time. There is definitely room for improving integration with dynamic formsets and fields that use javascript widgets, primarily on the `formset.js` side.

**Comment by rizumu** [comment]
@anneFly @codingjoe I've tried the standard `django-dynamic-formset`, but it breaks the original behaviour I had pre-rewrite. I still want to look into exactly why the fork I've listed above resolves that for me, but I just want to note for others that might hit similar issues to first dig into `formset.js`. `django-select2`, with the right `formset.js` behaves as it did pre-rewrite. Thanks again for all your help resolving this issue!
